File: tests/test_nodes/test_volatile.py
# Author: Nicolas Legrand <nicolas.legrand@cas.au.dk>

import numpy as np
from pyhgf.rshgf import Network as RsNetwork

from pyhgf import load_data
from pyhgf.model import Network as PyNetwork

NETWORK_CLASSES = [PyNetwork, RsNetwork]


def _assert_value_level_match(net_a, node_a, net_b, node_b, label=""):
    """Assert value-level trajectories match between two networks."""
    for key in ["mean", "expected_mean", "precision", "expected_precision"]:
        assert np.isclose(
            net_a.node_trajectories[node_a][key],
            net_b.node_trajectories[node_b][key],
        ).all(), f"{label}: Value-level key '{key}' mismatch"


def _assert_vol_level_match(volatile_net, vol_node, explicit_net, exp_node, label=""):
    """Assert volatility-level trajectories match (volatile _vol vs explicit node)."""
    vol_key_map = {
        "mean_vol": "mean",
        "expected_mean_vol": "expected_mean",
        "precision_vol": "precision",
        "expected_precision_vol": "expected_precision",
    }
    for vol_key, explicit_key in vol_key_map.items():
        assert np.isclose(
            volatile_net.node_trajectories[vol_node][vol_key],
            explicit_net.node_trajectories[exp_node][explicit_key],
        ).all(), (
            f"{label}: Volatility-level key '{vol_key}' vs '{explicit_key}' mismatch"
        )


def _run_volatile_vs_explicit(update_type):
    """Build volatile and explicit networks for each class, then cross-compare."""
    timeseries = load_data("continuous")

    volatile = {}
    explicit = {}
    for cls in NETWORK_CLASSES:
        name = cls.__name__
        volatile[name] = (
            cls(update_type=update_type)
            .add_nodes()
            .add_nodes(kind="volatile-state", value_children=0)
            .input_data(input_data=timeseries)
        )
        explicit[name] = (
            cls(update_type=update_type)
            .add_nodes()
            .add_nodes(value_children=0)
            .add_nodes(volatility_children=1)
            .input_data(input_data=timeseries)
        )

    # For each implementation: volatile must match explicit
    for name in volatile:
        label = f"{update_type} {name}"
        _assert_value_level_match(
            volatile[name], 0, explicit[name], 0, f"{label} input"
        )
        _assert_value_level_match(volatile[name], 1, explicit[name], 1, label)
        _assert_vol_level_match(volatile[name], 1, explicit[name], 2, label)

    # Cross-implementation: volatile networks must agree
    ref_name = NETWORK_CLASSES[0].__name__
    for name in volatile:
        if name == ref_name:
            continue
        label = f"{update_type} {ref_name} vs {name}"
        _assert_value_level_match(
            volatile[ref_name], 0, volatile[name], 0, f"{label} input"
        )
        _assert_value_level_match(volatile[ref_name], 1, volatile[name], 1, label)


def test_volatile_node_matches_explicit_volatility_parent():
    """Test volatile nodes against continuous pairs (standard update)."""
    _run_volatile_vs_explicit("standard")


def test_volatile_node_ehgf_matches_explicit():
    """Test volatile node with eHGF update matches explicit network."""
    _run_volatile_vs_explicit("eHGF")


def test_volatile_node_unbounded_matches_explicit():
    """Test volatile node with unbounded update matches explicit network."""
    _run_volatile_vs_explicit("unbounded")

File: tests/test_volatile.rs
use rshgf::model::Network;

/// Helper to assert approximate equality of f64 values.
fn assert_close(actual: f64, expected: f64, tol: f64, label: &str) {
    assert!(
        (actual - expected).abs() < tol,
        "{}: expected {}, got {} (diff = {})",
        label,
        expected,
        actual,
        (actual - expected).abs()
    );
}

/// Assert that the value-level trajectories of two nodes in two networks match.
fn assert_value_level_match(
    net_a: &Network,
    node_a: usize,
    net_b: &Network,
    node_b: usize,
    label: &str,
) {
    let floats_a = net_a
        .node_trajectories
        .floats
        .get(&node_a)
        .unwrap_or_else(|| panic!("{}: no trajectories for node {}", label, node_a));
    let floats_b = net_b
        .node_trajectories
        .floats
        .get(&node_b)
        .unwrap_or_else(|| panic!("{}: no trajectories for node {}", label, node_b));

    for key in ["mean", "expected_mean", "precision", "expected_precision"] {
        let a = &floats_a[key];
        let b = &floats_b[key];
        for (t, (va, vb)) in a.iter().zip(b.iter()).enumerate() {
            assert_close(
                *va,
                *vb,
                1e-6,
                &format!("{} value-level '{}' t={}", label, key, t),
            );
        }
    }
}

/// Assert that the volatility-level trajectories of a volatile node match the
/// corresponding explicit node trajectories.
fn assert_vol_level_match(
    volatile_net: &Network,
    vol_node: usize,
    explicit_net: &Network,
    exp_node: usize,
    label: &str,
) {
    let vol_floats = volatile_net
        .node_trajectories
        .floats
        .get(&vol_node)
        .unwrap_or_else(|| panic!("{}: no trajectories for volatile node {}", label, vol_node));
    let exp_floats = explicit_net
        .node_trajectories
        .floats
        .get(&exp_node)
        .unwrap_or_else(|| panic!("{}: no trajectories for explicit node {}", label, exp_node));

    let key_map = [
        ("mean_vol", "mean"),
        ("expected_mean_vol", "expected_mean"),
        ("precision_vol", "precision"),
        ("expected_precision_vol", "expected_precision"),
    ];

    for (vol_key, exp_key) in key_map {
        let a = &vol_floats[vol_key];
        let b = &exp_floats[exp_key];
        for (t, (va, vb)) in a.iter().zip(b.iter()).enumerate() {
            assert_close(
                *va,
                *vb,
                1e-6,
                &format!("{} vol-level '{}' vs '{}' t={}", label, vol_key, exp_key, t),
            );
        }
    }
}

/// Build a volatile network: input (node 0) + volatile-state value parent (node 1).
fn build_volatile_network(update_type: &str, data: &[f64]) -> Network {
    let mut net = Network::new(update_type);
    net.add_nodes("continuous-state", 1, None, None, None, None);
    net.add_nodes("volatile-state", 1, None, Some(0.into()), None, None);
    net.set_update_sequence();
    net.input_data(data.to_vec(), None);
    net
}

/// Build an explicit network: input (node 0) + value parent (node 1) + volatility
/// parent of node 1 (node 2).
fn build_explicit_network(update_type: &str, data: &[f64]) -> Network {
    let mut net = Network::new(update_type);
    net.add_nodes("continuous-state", 1, None, None, None, None);
    net.add_nodes("continuous-state", 1, None, Some(0.into()), None, None);
    net.add_nodes("continuous-state", 1, None, None, None, Some(1.into()));
    net.set_update_sequence();
    net.input_data(data.to_vec(), None);
    net
}

/// Run the volatile-vs-explicit comparison for the given update type.
fn compare_volatile_and_explicit(update_type: &str) {
    let data: Vec<f64> = (0..20).map(|i| (i as f64) * 0.1).collect();

    let volatile_net = build_volatile_network(update_type, &data);
    let explicit_net = build_explicit_network(update_type, &data);

    let label = format!("{} volatile vs explicit", update_type);

    // Input nodes should agree
    assert_value_level_match(&volatile_net, 0, &explicit_net, 0, &format!("{} input", label));

    // Value level of volatile node 1 should match explicit node 1
    assert_value_level_match(&volatile_net, 1, &explicit_net, 1, &label);

    // Volatility level of volatile node 1 should match explicit node 2
    assert_vol_level_match(&volatile_net, 1, &explicit_net, 2, &label);
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[test]
fn test_volatile_standard_matches_explicit() {
    compare_volatile_and_explicit("standard");
}

#[test]
fn test_volatile_ehgf_matches_explicit() {
    compare_volatile_and_explicit("eHGF");
}

#[test]
fn test_volatile_unbounded_matches_explicit() {
    compare_volatile_and_explicit("unbounded");
}
